Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/safetynet/SafetyNetClientWrapper.kt

```diff
@@ -0,0 +1,116 @@
+package de.rki.coronawarnapp.datadonation.safetynet
+
+import com.google.android.gms.common.api.ApiException
+import com.google.android.gms.common.api.CommonStatusCodes
+import com.google.android.gms.safetynet.SafetyNetApi
+import com.google.android.gms.safetynet.SafetyNetClient
+import com.google.gson.JsonObject
+import com.google.gson.JsonParser
+import dagger.Reusable
+import de.rki.coronawarnapp.datadonation.safetynet.SafetyNetException.Type
+import de.rki.coronawarnapp.environment.EnvironmentSetup
+import okio.ByteString.Companion.decodeBase64
+import timber.log.Timber
+import javax.inject.Inject
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+@Reusable
+class SafetyNetClientWrapper @Inject constructor(
+    private val safetyNetClient: SafetyNetClient,
+    private val environmentSetup: EnvironmentSetup
+) {
+
+    suspend fun attest(nonce: ByteArray): Report {
+        val jwsResult = callClient(nonce).jwsResult ?: throw SafetyNetException(
+            Type.ATTESTATION_FAILED,
+            "JWS was null"
+        )
+
+        val components = jwsResult.split(".")
+        if (components.size != 3) throw SafetyNetException(
+            Type.ATTESTATION_FAILED,
+            "Invalid JWS: Components are missing."
+        )
+
+        val header = try {
+            components[0].decodeBase64Json()
+        } catch (e: Exception) {
+            throw SafetyNetException(Type.ATTESTATION_FAILED, "Failed to decode JWS header.", e)
+        }
+        val body = try {
+            components[1].decodeBase64Json()
+        } catch (e: Exception) {
+            throw SafetyNetException(Type.ATTESTATION_FAILED, "Failed to decode JWS body.", e)
+        }
+
+        val signature = try {
+            components[2].decodeBase64()!!.toByteArray()
+        } catch (e: Exception) {
+            throw SafetyNetException(Type.ATTESTATION_FAILED, "Failed to decode JWS signature.", e)
+        }
+
+        return Report(
+            jwsResult = jwsResult,
+            header = header,
+            body = body,
+            signature = signature
+        )
+    }
+
+    private fun String.decodeBase64Json(): JsonObject {
+        val rawJson = decodeBase64()!!.string(Charsets.UTF_8)
+        return JsonParser.parseString(rawJson).asJsonObject
+    }
+
+    private suspend fun callClient(nonce: ByteArray): SafetyNetApi.AttestationResponse = suspendCoroutine { cont ->
```

Comment:
possibly a timeout would be helpful here...

Reply:
According to the [documentation](https://developer.android.com/training/safetynet/attestation#verify-attestation-response), only for the blocking version of the request (we are using the non-blocking version), a timeout is necessary:

> Note: The preceding code snippet uses a non-blocking version of the attest() method. If you're making a blocking call instead, always set a response timeout. That way, your app remains responsive, even if a slow network connection causes a delayed response from the SafetyNet Attestation API.

Reply:
A timeout is useful, but as survey and analytics are very different use-cases I'd prefer if the timeout is handled downstream by the caller. What do you think?

Reply:
I've added a timeout of 30 seconds for the attestation call as "maximum", lower timeout values can be enforced by the caller.